In [66]:
import pandas as pd

In [67]:
fighters = pd.read_csv(
    "../data/raw/kaggle_ufc_1994_2026/ufc_fighters_final.csv"
)

fights = pd.read_csv(
    "../data/raw/kaggle_ufc_1994_2026/ufc_gold_dataset_final.csv"
)

In [68]:
print("Fighters shape:", fighters.shape)
print("Fights shape:", fights.shape)

Fighters shape: (4455, 18)
Fights shape: (8551, 37)


In [69]:
fighters.head()

,Fighter_Name,Height,Weight,Reach,Stance,DOB,Wins,Losses,Draws,SLpM,Str_Acc,SApM,Str_Def,TD_Avg,TD_Acc,TD_Def,Sub_Avg,Fighter_URL
0,Tom Aaron,NaN,155 lbs.,NaN,NaN,1978-07-13,5,3,0,0.00,0%,0.00,0%,0.00,0%,0%,0.0,http://ufcstats.com/fighter-details/93fe7332d1...
1,Danny Abbadi,"5' 11""",155 lbs.,NaN,Orthodox,1983-07-03,4,6,0,3.29,38%,4.41,57%,0.00,0%,77%,0.0,http://ufcstats.com/fighter-details/15df64c02b...
2,Nariman Abbasov,"5' 8""",155 lbs.,"66.0""",Orthodox,1994-02-01,28,4,0,3.00,20%,5.67,46%,0.00,0%,66%,0.0,http://ufcstats.com/fighter-details/59a9d6dac6...
3,Darion Abbey,"6' 2""",265 lbs.,"80.0""",Orthodox,1993-02-25,9,5,0,8.44,50%,14.06,28%,0.00,0%,0%,0.0,http://ufcstats.com/fighter-details/4961467134...
4,David Abbott,"6' 0""",265 lbs.,NaN,Switch,1965-04-26,10,15,0,1.35,30%,3.55,38%,1.07,33%,66%,0.0,http://ufcstats.com/fighter-details/b361180739...


In [70]:
fights.head()

,Fight_URL,Fighter_1,Fighter_2,Winner,Weight_Class,Method,End_Round,End_Time,Total_Fight_Time_Sec,Time_Format,...,F2_Body,F1_Leg,F2_Leg,F1_Distance,F2_Distance,F1_Clinch,F2_Clinch,F1_Ground,F2_Ground,Event_Date
0,http://ufcstats.com/fight-details/4acab67848e7...,Scott Morris,Sean Daugherty,Scott Morris,Open Weight Bout,Submission,1,0:20,20,No Time Limit,...,0,0,0,0,0,1,0,0,0,1994-03-11
1,http://ufcstats.com/fight-details/4b9ae533ccb3...,Patrick Smith,Ray Wizard,Patrick Smith,Open Weight Bout,Submission,1,0:58,58,No Time Limit,...,0,0,1,0,1,1,0,0,0,1994-03-11
2,http://ufcstats.com/fight-details/00835554f95f...,Royce Gracie,Patrick Smith,Royce Gracie,UFC 2 Tournament Title Bout,KO/TKO,1,1:17,77,No Time Limit,...,1,1,0,0,0,1,1,3,0,1994-03-11
3,http://ufcstats.com/fight-details/aa161c715974...,Royce Gracie,Remco Pardoel,Royce Gracie,Open Weight Bout,Submission,1,1:31,91,No Time Limit,...,0,0,0,0,0,0,0,0,0,1994-03-11
4,http://ufcstats.com/fight-details/8fbdde20b712...,Patrick Smith,Johnny Rhodes,Patrick Smith,Open Weight Bout,Submission,1,1:07,67,No Time Limit,...,0,2,2,3,4,2,0,0,0,1994-03-11


In [71]:
print("Fighter columns:")
print(fighters.columns.tolist())

print("\nFight columns:")
print(fights.columns.tolist())

Fighter columns:
['Fighter_Name', 'Height', 'Weight', 'Reach', 'Stance', 'DOB', 'Wins', 'Losses', 'Draws', 'SLpM', 'Str_Acc', 'SApM', 'Str_Def', 'TD_Avg', 'TD_Acc', 'TD_Def', 'Sub_Avg', 'Fighter_URL']

Fight columns:
['Fight_URL', 'Fighter_1', 'Fighter_2', 'Winner', 'Weight_Class', 'Method', 'End_Round', 'End_Time', 'Total_Fight_Time_Sec', 'Time_Format', 'F1_KD', 'F2_KD', 'F1_Sig_Landed', 'F1_Sig_Att', 'F2_Sig_Landed', 'F2_Sig_Att', 'F1_TD_Landed', 'F2_TD_Landed', 'F1_TD_Att', 'F2_TD_Att', 'F1_Sub_Att', 'F2_Sub_Att', 'F1_Ctrl_Sec', 'F2_Ctrl_Sec', 'F1_Head', 'F2_Head', 'F1_Body', 'F2_Body', 'F1_Leg', 'F2_Leg', 'F1_Distance', 'F2_Distance', 'F1_Clinch', 'F2_Clinch', 'F1_Ground', 'F2_Ground', 'Event_Date']


In [72]:
fighters.info()

<class 'pandas.DataFrame'>
RangeIndex: 4455 entries, 0 to 4454
Data columns (total 18 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Fighter_Name  4455 non-null   str    
 1   Height        4137 non-null   str    
 2   Weight        4369 non-null   str    
 3   Reach         2515 non-null   str    
 4   Stance        3606 non-null   str    
 5   DOB           3949 non-null   str    
 6   Wins          4455 non-null   int64  
 7   Losses        4455 non-null   int64  
 8   Draws         4455 non-null   int64  
 9   SLpM          4455 non-null   float64
 10  Str_Acc       4455 non-null   str    
 11  SApM          4455 non-null   float64
 12  Str_Def       4455 non-null   str    
 13  TD_Avg        4455 non-null   float64
 14  TD_Acc        4455 non-null   str    
 15  TD_Def        4455 non-null   str    
 16  Sub_Avg       4455 non-null   float64
 17  Fighter_URL   4455 non-null   str    
dtypes: float64(4), int64(3), str(11)
memory

In [73]:
fights.info()

<class 'pandas.DataFrame'>
RangeIndex: 8551 entries, 0 to 8550
Data columns (total 37 columns):
 #   Column                Non-Null Count  Dtype
---  ------                --------------  -----
 0   Fight_URL             8551 non-null   str  
 1   Fighter_1             8551 non-null   str  
 2   Fighter_2             8551 non-null   str  
 3   Winner                8551 non-null   str  
 4   Weight_Class          8551 non-null   str  
 5   Method                8551 non-null   str  
 6   End_Round             8551 non-null   int64
 7   End_Time              8551 non-null   str  
 8   Total_Fight_Time_Sec  8551 non-null   int64
 9   Time_Format           8551 non-null   str  
 10  F1_KD                 8551 non-null   int64
 11  F2_KD                 8551 non-null   int64
 12  F1_Sig_Landed         8551 non-null   int64
 13  F1_Sig_Att            8551 non-null   int64
 14  F2_Sig_Landed         8551 non-null   int64
 15  F2_Sig_Att            8551 non-null   int64
 16  F1_TD_Landed     

In [74]:
fighter_missing = pd.DataFrame({
    "missing_count": fighters.isna().sum(),
    "missing_percent": fighters.isna().mean() * 100
})

fighter_missing.sort_values("missing_percent", ascending=False)

,missing_count,missing_percent
Reach,1940,43.546577
Stance,849,19.057239
DOB,506,11.358025
Height,318,7.138047
Weight,86,1.930415
Fighter_Name,0,0.000000
Str_Def,0,0.000000
Sub_Avg,0,0.000000
TD_Def,0,0.000000
TD_Acc,0,0.000000


In [75]:
numeric_stat_columns = [
    "SLpM",
    "SApM",
    "TD_Avg",
    "Sub_Avg",
]

percent_stat_columns = [
    "Str_Acc",
    "Str_Def",
    "TD_Acc",
    "TD_Def",
]

zero_counts = {}

for column in numeric_stat_columns:
    zero_counts[column] = (fighters[column] == 0).sum()

for column in percent_stat_columns:
    zero_counts[column] = (fighters[column] == "0%").sum()

zero_audit = pd.DataFrame.from_dict(
    zero_counts,
    orient="index",
    columns=["zero_count"],
)

zero_audit["zero_percent"] = (
    zero_audit["zero_count"] / len(fighters) * 100
)

zero_audit.sort_values("zero_percent", ascending=False)

,zero_count,zero_percent
Sub_Avg,2516,56.475870
TD_Avg,1856,41.661055
TD_Acc,1856,41.661055
TD_Def,1549,34.769921
SLpM,801,17.979798
Str_Acc,801,17.979798
Str_Def,734,16.475870
SApM,724,16.251403


In [76]:
fight_audit = fights.copy()

fight_audit["Event_Date"] = pd.to_datetime(
    fight_audit["Event_Date"]
)

fight_audit["Year"] = fight_audit["Event_Date"].dt.year

core_stat_columns = [
    "F1_KD",
    "F2_KD",
    "F1_Sig_Landed",
    "F1_Sig_Att",
    "F2_Sig_Landed",
    "F2_Sig_Att",
    "F1_TD_Landed",
    "F2_TD_Landed",
    "F1_TD_Att",
    "F2_TD_Att",
    "F1_Sub_Att",
    "F2_Sub_Att",
    "F1_Ctrl_Sec",
    "F2_Ctrl_Sec",
]

fight_audit["all_core_stats_zero"] = (
    fight_audit[core_stat_columns] == 0
).all(axis=1)

zero_stats_by_year = (
    fight_audit
    .groupby("Year")
    .agg(
        total_fights=("Fight_URL", "count"),
        all_zero_fights=("all_core_stats_zero", "sum"),
    )
)

zero_stats_by_year["zero_percent"] = (
    zero_stats_by_year["all_zero_fights"]
    / zero_stats_by_year["total_fights"]
    * 100
)

zero_stats_by_year

,total_fights,all_zero_fights,zero_percent
Year,,,
1994,29,0,0.0
1995,34,0,0.0
1996,36,0,0.0
1997,39,0,0.0
1998,21,0,0.0
1999,44,0,0.0
2000,43,0,0.0
2001,40,0,0.0
2002,53,0,0.0


In [77]:
fight_audit["no_sig_strike_stats"] = (
    (fight_audit["F1_Sig_Att"] == 0)
    & (fight_audit["F2_Sig_Att"] == 0)
)

fight_audit["no_takedown_stats"] = (
    (fight_audit["F1_TD_Att"] == 0)
    & (fight_audit["F2_TD_Att"] == 0)
)

fight_audit["no_control_stats"] = (
    (fight_audit["F1_Ctrl_Sec"] == 0)
    & (fight_audit["F2_Ctrl_Sec"] == 0)
)

stats_by_year = (
    fight_audit
    .groupby("Year")
    .agg(
        total_fights=("Fight_URL", "count"),
        no_sig_strike_stats=("no_sig_strike_stats", "sum"),
        no_takedown_stats=("no_takedown_stats", "sum"),
        no_control_stats=("no_control_stats", "sum"),
    )
)

stats_by_year["no_sig_strike_percent"] = (
    stats_by_year["no_sig_strike_stats"]
    / stats_by_year["total_fights"]
    * 100
)

stats_by_year["no_takedown_percent"] = (
    stats_by_year["no_takedown_stats"]
    / stats_by_year["total_fights"]
    * 100
)

stats_by_year["no_control_percent"] = (
    stats_by_year["no_control_stats"]
    / stats_by_year["total_fights"]
    * 100
)

stats_by_year

,total_fights,no_sig_strike_stats,no_takedown_stats,no_control_stats,no_sig_strike_percent,no_takedown_percent,no_control_percent
Year,,,,,,,
1994,29,2,2,29,6.896552,6.896552,100.000000
1995,34,2,5,34,5.882353,14.705882,100.000000
1996,36,0,3,36,0.000000,8.333333,100.000000
1997,39,0,5,39,0.000000,12.820513,100.000000
1998,21,0,1,21,0.000000,4.761905,100.000000
1999,44,0,5,22,0.000000,11.363636,50.000000
2000,43,1,1,1,2.325581,2.325581,2.325581
2001,40,0,4,0,0.000000,10.000000,0.000000
2002,53,0,10,3,0.000000,18.867925,5.660377


In [78]:
fight_audit["has_any_control"] = (
    (fight_audit["F1_Ctrl_Sec"] > 0)
    | (fight_audit["F2_Ctrl_Sec"] > 0)
)

control_by_event = (
    fight_audit
    .groupby("Event_Date")
    .agg(
        total_fights=("Fight_URL", "count"),
        fights_with_control=("has_any_control", "sum"),
    )
    .reset_index()
)

control_by_event["control_signal_percent"] = (
    control_by_event["fights_with_control"]
    / control_by_event["total_fights"]
    * 100
)

In [79]:
control_by_event[
    (control_by_event["Event_Date"] >= "1998-01-01")
    & (control_by_event["Event_Date"] <= "2001-12-31")
]

,Event_Date,total_fights,fights_with_control,control_signal_percent
17,1998-03-13,6,0,0.000000
18,1998-05-15,8,0,0.000000
19,1998-10-16,7,0,0.000000
20,1999-01-08,7,0,0.000000
21,1999-03-05,7,0,0.000000
22,1999-05-07,7,0,0.000000
23,1999-07-16,8,8,100.000000
24,1999-09-24,8,7,87.500000
25,1999-11-19,7,7,100.000000
26,2000-03-10,7,7,100.000000


### Provisional finding: control-time coverage

Control-time values appear to have a structural coverage boundary around
July 16, 1999 (UFC 21). Earlier fights frequently contain zeros for both
fighters, which may represent unavailable data rather than true zero control.
This requires source validation before cleaning.

**Evidence:** All 7 fights from UFC 20 (1999-05-07) have zero control
time for both fighters, while all 8 fights from the immediately following
UFC 21 (1999-07-16) contain a positive control-time value for at least one
fighter.

**Working assumption:** Pre-UFC-21 control-time zeros represent unavailable
data rather than true zero control. Do not modify raw data; handle this later
during normalization/cleaning.

In [80]:
boundary_samples = fight_audit[
    fight_audit["Event_Date"].isin(
        [
            pd.Timestamp("1999-05-07"),
            pd.Timestamp("1999-07-16"),
        ]
    )
][
    [
        "Event_Date",
        "Fighter_1",
        "Fighter_2",
        "F1_Ctrl_Sec",
        "F2_Ctrl_Sec",
        "Fight_URL",
    ]
]

boundary_samples

,Event_Date,Fighter_1,Fighter_2,F1_Ctrl_Sec,F2_Ctrl_Sec,Fight_URL
173,1999-05-07,Ron Waterman,Chris Condo,0,0,http://ufcstats.com/fight-details/79f9086932f2...
174,1999-05-07,Laverne Clark,Fabiano Iha,0,0,http://ufcstats.com/fight-details/3a271d9d9210...
175,1999-05-07,Marcelo Mello,David Roberts,0,0,http://ufcstats.com/fight-details/6bb2a427004d...
176,1999-05-07,Wanderlei Silva,Tony Petarra,0,0,http://ufcstats.com/fight-details/c892b23ad8f2...
177,1999-05-07,Pete Williams,Travis Fulton,0,0,http://ufcstats.com/fight-details/a3fe8217502c...
178,1999-05-07,Pedro Rizzo,Tra Telligman,0,0,http://ufcstats.com/fight-details/99b5010fd259...
179,1999-05-07,Bas Rutten,Kevin Randleman,0,0,http://ufcstats.com/fight-details/73c07530f667...
180,1999-07-16,Maurice Smith,Marco Ruas,34,221,http://ufcstats.com/fight-details/62ac5bff3c18...
181,1999-07-16,Pat Miletich,Andre Pederneiras,96,4,http://ufcstats.com/fight-details/df8d4014d820...
182,1999-07-16,Jeremy Horn,Daiju Takase,178,0,http://ufcstats.com/fight-details/38e3986b7cff...


In [81]:
fighters["Fighter_URL"].duplicated().sum()

np.int64(0)

In [82]:
fights["Fight_URL"].duplicated().sum()

np.int64(0)

In [83]:
print("Earliest fight:", fight_audit["Event_Date"].min())
print("Latest fight:", fight_audit["Event_Date"].max())

Earliest fight: 1994-03-11 00:00:00
Latest fight: 2026-03-07 00:00:00


In [84]:
fight_fighters = set(fights["Fighter_1"]) | set(fights["Fighter_2"])
profile_fighters = set(fighters["Fighter_Name"])

missing_profiles = sorted(fight_fighters - profile_fighters)

print("Unique fighters appearing in fights:", len(fight_fighters))
print("Fighters in profile table:", len(profile_fighters))
print("Fight fighters missing a profile:", len(missing_profiles))

Unique fighters appearing in fights: 2647
Fighters in profile table: 4448
Fight fighters missing a profile: 0


In [85]:
missing_profiles[:20]

[]

In [86]:
duplicate_names = (
    fighters["Fighter_Name"]
    .value_counts()
    .loc[lambda counts: counts > 1]
)

duplicate_names

Fighter_Name
Mike Davis           2
Joey Gomez           2
Tony Johnson         2
Michael McDonald     2
Jean Silva           2
Bruno Silva          2
Victor Valenzuela    2
Name: count, dtype: int64

In [87]:
fighters[
    fighters["Fighter_Name"].isin(duplicate_names.index)
][
    [
        "Fighter_Name",
        "DOB",
        "Height",
        "Weight",
        "Fighter_URL",
    ]
].sort_values("Fighter_Name")

,Fighter_Name,DOB,Height,Weight,Fighter_URL
3698,Bruno Silva,1990-03-16,"5' 4""",125 lbs.,http://ufcstats.com/fighter-details/294aa73dbf...
3699,Bruno Silva,1989-07-13,"6' 0""",185 lbs.,http://ufcstats.com/fighter-details/12ebd7d157...
3684,Jean Silva,1977-10-08,"5' 6""",160 lbs.,http://ufcstats.com/fighter-details/9211aae062...
3709,Jean Silva,1996-12-27,"5' 7""",145 lbs.,http://ufcstats.com/fighter-details/52ef95b586...
1414,Joey Gomez,1986-07-21,"5' 10""",135 lbs.,http://ufcstats.com/fighter-details/0778f94eb5...
1416,Joey Gomez,1989-08-29,"5' 10""",155 lbs.,http://ufcstats.com/fighter-details/3a28e1e641...
2513,Michael McDonald,1965-02-06,"5' 11""",205 lbs.,http://ufcstats.com/fighter-details/d52ef69410...
2515,Michael McDonald,1991-01-15,"5' 9""",135 lbs.,http://ufcstats.com/fighter-details/d0314416a7...
882,Mike Davis,NaN,NaN,NaN,http://ufcstats.com/fighter-details/c8661e204c...
887,Mike Davis,1992-10-07,"6' 0""",155 lbs.,http://ufcstats.com/fighter-details/fb3e61720b...


In [88]:
profile_only_fighters = sorted(
    profile_fighters - fight_fighters
)

print("Profile-only fighters:", len(profile_only_fighters))

profile_only_fighters[:30]

Profile-only fighters: 1801


['AJ Fonseca',
 'AJ Matthews',
 'AJ McKee',
 'AJ Siscoe',
 'Aaron Ely',
 'Aaron Jeffery',
 'Aaron Lanfranco',
 'Aaron Miller',
 'Aaron Tau',
 'Aaron Trujillo',
 'Aaron Wetherspoon',
 'Abdellah Er-Ramy',
 'Abdul Azeem Badakhshi',
 'Abel Cullum',
 'Abongo Humphrey',
 'Abram Torres',
 'Acacio Dos Santos',
 'Achilles Estremadura',
 'Achmed Labasanov',
 'Adam Antolin',
 'Adam Bramhald',
 'Adam Hunter',
 'Adam Khaliev',
 'Adam Livingston',
 'Adam Lynn',
 'Adam Meredith',
 'Adam Piccolotti',
 'Adam Schindler',
 'Adam Smith',
 'Adam Steele']

In [89]:
duplicate_name_list = set(duplicate_names.index)

fight_rows_with_duplicate_names = fights[
    fights["Fighter_1"].isin(duplicate_name_list)
    | fights["Fighter_2"].isin(duplicate_name_list)
][
    [
        "Event_Date",
        "Fighter_1",
        "Fighter_2",
        "Weight_Class",
        "Fight_URL",
    ]
].sort_values("Event_Date")

fight_rows_with_duplicate_names

,Event_Date,Fighter_1,Fighter_2,Weight_Class,Fight_URL
1576,2011-03-26,Michael McDonald,Edwin Figueroa,Bantamweight Bout,http://ufcstats.com/fight-details/c67473192bda...
1597,2011-05-28,Michael McDonald,Chris Cariaso,Bantamweight Bout,http://ufcstats.com/fight-details/556952a9520d...
1755,2011-11-19,Michael McDonald,Alex Soto,Bantamweight Bout,http://ufcstats.com/fight-details/52f77a9f9eb6...
1884,2012-04-21,Miguel Torres,Michael McDonald,Bantamweight Bout,http://ufcstats.com/fight-details/33a93002fef6...
2171,2013-02-16,Renan Barao,Michael McDonald,UFC Interim Bantamweight Title Bout,http://ufcstats.com/fight-details/3dbf17b9946d...
2349,2013-08-17,Brad Pickett,Michael McDonald,Bantamweight Bout,http://ufcstats.com/fight-details/8e41d022b552...
2508,2013-12-14,Urijah Faber,Michael McDonald,Bantamweight Bout,http://ufcstats.com/fight-details/00566b4d8956...
3504,2016-01-02,Michael McDonald,Masanori Kanehara,Bantamweight Bout,http://ufcstats.com/fight-details/a19d619f1b9b...
3522,2016-01-17,Rob Font,Joey Gomez,Bantamweight Bout,http://ufcstats.com/fight-details/18cdc16fc7ab...
3735,2016-07-13,Michael McDonald,John Lineker,Bantamweight Bout,http://ufcstats.com/fight-details/08c6b88ac227...


### Fighter identity finding

`Fighter_Name` is not globally unique. Seven names occur for multiple
distinct fighter profiles, each with a unique `Fighter_URL`.

Some ambiguous names also appear in the fight table, meaning fighter names
cannot safely serve as permanent join keys.

UPSET should eventually use its own internal fighter ID and maintain
source-specific IDs for reconciliation across data providers.

## Profile-Only Fighter Investigation

The fighter profile dataset contains 1,801 names that do not appear in the historical fight dataset.

The goal of this section is not to delete these fighters. We want to briefly investigate why the profile dataset has broader coverage than the fight dataset and determine what role these records should have in UPSET.

In [90]:
# Collect every fighter name that appears in the historical fight table.
fight_fighter_names = set(fights["Fighter_1"].dropna()) | set(
    fights["Fighter_2"].dropna()
)

# Select fighter profiles whose names never appear in either fight column.
profile_only = fighters[
    ~fighters["Fighter_Name"].isin(fight_fighter_names)
].copy()

print(f"Profile-only fighter rows: {len(profile_only):,}")
print()

profile_only[
    [
        "Fighter_Name",
        "Weight",
        "Wins",
        "Losses",
        "Draws",
        "Fighter_URL",
    ]
].head(25)

Profile-only fighter rows: 1,803



,Fighter_Name,Weight,Wins,Losses,Draws,Fighter_URL
0,Tom Aaron,155 lbs.,5,3,0,http://ufcstats.com/fighter-details/93fe7332d1...
2,Nariman Abbasov,155 lbs.,28,4,0,http://ufcstats.com/fighter-details/59a9d6dac6...
3,Darion Abbey,265 lbs.,9,5,0,http://ufcstats.com/fighter-details/4961467134...
8,Hiroyuki Abe,145 lbs.,8,15,3,http://ufcstats.com/fighter-details/c0ed7b2081...
13,Cyborg Abreu,NaN,0,0,0,http://ufcstats.com/fighter-details/f689bd7bbd...
14,Daniel Acacio,180 lbs.,30,18,0,http://ufcstats.com/fighter-details/1c5879330d...
15,John Adajar,170 lbs.,6,2,0,http://ufcstats.com/fighter-details/989b85f654...
18,Anthony Adams,185 lbs.,8,2,0,http://ufcstats.com/fighter-details/a77633a989...
22,Mohamed Ado,170 lbs.,5,1,0,http://ufcstats.com/fighter-details/7a846fcbf1...
25,Kantharaj Agasa,135 lbs.,12,3,0,http://ufcstats.com/fighter-details/26f4288e83...


In [91]:
profile_only["Fighter_URL"].str.extract(
    r"https?://([^/]+)",
    expand=False,
).value_counts()

Fighter_URL
ufcstats.com    1803
Name: count, dtype: int64

In [92]:
print(f"Profile-only rows: {len(profile_only):,}")
print(f"Profile-only unique names: {profile_only['Fighter_Name'].nunique():,}")
print()

profile_only[
    profile_only["Fighter_Name"].duplicated(keep=False)
][
    [
        "Fighter_Name",
        "DOB",
        "Height",
        "Weight",
        "Fighter_URL",
    ]
].sort_values("Fighter_Name")

Profile-only rows: 1,803
Profile-only unique names: 1,801



,Fighter_Name,DOB,Height,Weight,Fighter_URL
1874,Tony Johnson,1983-05-02,"6' 2""",205 lbs.,http://ufcstats.com/fighter-details/3641a0d117...
1882,Tony Johnson,NaN,"6' 1""",265 lbs.,http://ufcstats.com/fighter-details/a45bab4995...
4120,Victor Valenzuela,NaN,"5' 10""",155 lbs.,http://ufcstats.com/fighter-details/de277a4abc...
4122,Victor Valenzuela,1994-02-09,"5' 9""",170 lbs.,http://ufcstats.com/fighter-details/078695e385...


### Profile-Only Fighter Conclusion

The fighter profile dataset contains 1,803 profile-only rows representing 1,801 unique names.

The row/name difference is caused by two duplicate-name identity pairs:
- Tony Johnson
- Victor Valenzuela

All profile-only records use `ufcstats.com` fighter URLs.

External validation suggests that UFCStats' fighter directory has broader scope than the historical UFC fight dataset. Some profileonly fighters have records from UFC adjacent or other promotions such as DWCS, Strikeforce, and ProElite.

Therefore, profile-only fighters should not be treated as invalid records or deleted.

Working conclusion:

- the fighter profile table has broader coverage than the fight table
- the fight table should define the historical UFC fight universe for this dataset
- fighter profiles may still be useful for identity resolution, enrichment, and future analysis
- filtering should happen for a specific analytical purpose rather than by deleting profile-only fighters globally

## Cito API Validation

We are beginning evaluation of Cito as a secondary/current UFC data source.

The first validation step compares one fighter available in both Cito and the historical Kaggle fighter table.

We are using Islam Makhachev as the initial overlap example.

In [93]:
islam_kaggle = fighters[
    fighters["Fighter_Name"] == "Islam Makhachev"
]

islam_kaggle[
    [
        "Fighter_Name",
        "Height",
        "Weight",
        "Reach",
        "Stance",
        "Wins",
        "Losses",
        "Draws",
        "SLpM",
        "SApM",
        "TD_Avg",
        "TD_Acc",
        "TD_Def",
        "Sub_Avg",
        "Fighter_URL",
    ]
]

,Fighter_Name,Height,Weight,Reach,Stance,Wins,Losses,Draws,SLpM,SApM,TD_Avg,TD_Acc,TD_Def,Sub_Avg,Fighter_URL
2387,Islam Makhachev,"5' 10""",170 lbs.,"70.0""",Southpaw,28,1,0,2.45,1.45,3.1,56%,91%,1.0,http://ufcstats.com/fighter-details/275aca31f6...


In [94]:
cito_islam = {
    "Fighter_Name": "Islam Makhachev",
    "Height": '5\' 10"',
    "Weight": "170 lbs.",
    "Reach": '70.5"',
    "Stance": "Southpaw",
    "Wins": 29,
    "Losses": 1,
    "Draws": 0,
    "SLpM": 2.27,
    "SApM": 1.42,
    "TD_Avg": 3.22,
    "TD_Acc": 54.55,
    "TD_Def": 91,
    "Sub_Avg": 0.87,
}

kaggle_islam = islam_kaggle.iloc[0]

comparison = pd.DataFrame({
    "Kaggle": {
        "Height": kaggle_islam["Height"],
        "Weight": kaggle_islam["Weight"],
        "Reach": kaggle_islam["Reach"],
        "Stance": kaggle_islam["Stance"],
        "Wins": kaggle_islam["Wins"],
        "Losses": kaggle_islam["Losses"],
        "SLpM": kaggle_islam["SLpM"],
        "SApM": kaggle_islam["SApM"],
        "TD_Avg": kaggle_islam["TD_Avg"],
        "TD_Acc": kaggle_islam["TD_Acc"],
        "TD_Def": kaggle_islam["TD_Def"],
        "Sub_Avg": kaggle_islam["Sub_Avg"],
    },
    "Cito": {
        "Height": cito_islam["Height"],
        "Weight": cito_islam["Weight"],
        "Reach": cito_islam["Reach"],
        "Stance": cito_islam["Stance"],
        "Wins": cito_islam["Wins"],
        "Losses": cito_islam["Losses"],
        "SLpM": cito_islam["SLpM"],
        "SApM": cito_islam["SApM"],
        "TD_Avg": cito_islam["TD_Avg"],
        "TD_Acc": f"{cito_islam['TD_Acc']}%",
        "TD_Def": f"{cito_islam['TD_Def']}%",
        "Sub_Avg": cito_islam["Sub_Avg"],
    },
})

comparison

,Kaggle,Cito
Height,"5' 10""","5' 10"""
Weight,170 lbs.,170 lbs.
Reach,"70.0""","70.5"""
Stance,Southpaw,Southpaw
Wins,28,29
Losses,1,1
SLpM,2.45,2.27
SApM,1.45,1.42
TD_Avg,3.1,3.22
TD_Acc,56%,54.55%


In [95]:
import os
from pathlib import Path

import requests
from dotenv import load_dotenv

# Find the project root whether VS Code runs the notebook
# from /upset or /upset/notebooks.
current_dir = Path.cwd()
project_root = (
    current_dir.parent
    if current_dir.name == "notebooks"
    else current_dir
)

# Load secrets from the local .env file.
load_dotenv(project_root / ".env")

cito_api_key = os.getenv("CITO_API_KEY")

if not cito_api_key:
    raise RuntimeError("CITO_API_KEY was not found in .env")

url = "https://api.citoapi.com/api/v1/ufc/search"

response = requests.get(
    url,
    params={"q": "islam"},
    headers={"x-api-key": cito_api_key},
    timeout=30,
)

print("HTTP status:", response.status_code)

HTTP status: 200


In [96]:
# Convert the JSON response into normal Python objects.
cito_data = response.json()

# The search endpoint can return multiple fighters.
cito_fighters = cito_data["data"]["fighters"]

print(f"Fighters returned: {len(cito_fighters)}")

for fighter in cito_fighters:
    print(fighter["name"])

Fighters returned: 2
Islam Makhachev
Islam Dulatov


In [97]:
cito_islam = next(
    fighter
    for fighter in cito_fighters
    if fighter["name"] == "Islam Makhachev"
)

print("ID:", cito_islam["id"])
print("Name:", cito_islam["name"])
print("Record:", cito_islam["recordText"])
print("Reach:", cito_islam["reachInches"])

ID: 2df4f188-a33e-463a-a66b-471cfa23e2a0
Name: Islam Makhachev
Record: 29-1-0 (W-L-D)
Reach: 70.5


In [98]:
cito_stats = cito_islam["stats"]

comparison_live = pd.DataFrame({
    "Kaggle": {
        "Height": kaggle_islam["Height"],
        "Weight": kaggle_islam["Weight"],
        "Reach": kaggle_islam["Reach"],
        "Stance": kaggle_islam["Stance"],
        "Wins": kaggle_islam["Wins"],
        "Losses": kaggle_islam["Losses"],
        "SLpM": kaggle_islam["SLpM"],
        "SApM": kaggle_islam["SApM"],
        "TD_Avg": kaggle_islam["TD_Avg"],
        "TD_Acc": kaggle_islam["TD_Acc"],
        "TD_Def": kaggle_islam["TD_Def"],
        "Sub_Avg": kaggle_islam["Sub_Avg"],
    },
    "Cito": {
        "Height": f'{int(cito_islam["heightInches"]) // 12}\' '
                  f'{int(cito_islam["heightInches"]) % 12}"',
        "Weight": f'{cito_islam["weightLbs"]} lbs.',
        "Reach": f'{cito_islam["reachInches"]}"',
        "Stance": cito_islam["stance"],
        "Wins": cito_islam["recordWins"],
        "Losses": cito_islam["recordLosses"],
        "SLpM": cito_stats["sigStrikesLandedPerMin"],
        "SApM": cito_stats["sigStrikesAbsorbedPerMin"],
        "TD_Avg": cito_stats["takedownAvgPer15Min"],
        "TD_Acc": f'{float(cito_stats["takedownAccuracy"]) * 100:.2f}%',
        "TD_Def": f'{float(cito_stats["takedownDefense"]) * 100:.0f}%',
        "Sub_Avg": cito_stats["submissionAvgPer15Min"],
    },
})

comparison_live

,Kaggle,Cito
Height,"5' 10""","5' 10"""
Weight,170 lbs.,170 lbs.
Reach,"70.0""","70.5"""
Stance,Southpaw,Southpaw
Wins,28,29
Losses,1,1
SLpM,2.45,2.27
SApM,1.45,1.42
TD_Avg,3.1,3.22
TD_Acc,56%,54.55%


In [99]:
# Find the UFC 302 bout from the search response we already downloaded.
poirier_bout = next(
    bout
    for bout in cito_data["data"]["bouts"]
    if bout["eventSlug"] == "ufc-302-makhachev-vs-poirier"
)

bout_id = poirier_bout["id"]

print("Bout ID:", bout_id)
print(
    "Fight:",
    poirier_bout["fighters"][0]["fighterName"],
    "vs",
    poirier_bout["fighters"][1]["fighterName"],
)
print("Result:", poirier_bout["method"])
print("Finish:", f'Round {poirier_bout["resultRound"]} at {poirier_bout["resultTime"]}')

Bout ID: 1a0031e2072f21b6
Fight: Dustin Poirier vs Islam Makhachev
Result: SUB
Finish: Round 5 at 2:42


In [100]:
rounds_url = (
    f"https://api.citoapi.com/api/v1/ufc/bouts/{bout_id}/rounds"
)

rounds_response = requests.get(
    rounds_url,
    headers={"x-api-key": cito_api_key},
    timeout=30,
)

print("HTTP status:", rounds_response.status_code)

HTTP status: 403


In [101]:
recent_response = requests.get(
    "https://api.citoapi.com/api/v1/ufc/events/recent",
    headers={"x-api-key": cito_api_key},
    timeout=30,
)

print("HTTP status:", recent_response.status_code)

recent_data = recent_response.json()

for event in recent_data["data"][:5]:
    print(
        event.get("eventDate"),
        "-",
        event.get("title"),
        "- hasStats:",
        event.get("hasStats"),
    )

HTTP status: 200
2026-09-15 - Dana White's Contender Series 10.6 - hasStats: False
2026-09-12 - Noche UFC - hasStats: True
2026-09-08 - Dana White's Contender Series 10.5 - hasStats: False
2026-09-05 - UFC Fight Night: Hooker vs Parnasse - hasStats: True
2026-09-01 - Dana White's Contender Series 10.4 - hasStats: False


In [102]:
recent_event = next(
    event
    for event in recent_data["data"]
    if event.get("hasStats")
)

print("Event:", recent_event["title"])
print("Date:", recent_event["eventDate"])
print("Slug:", recent_event["slug"])

Event: Noche UFC
Date: 2026-09-12
Slug: ufc-fight-night-september-12-2026


In [103]:
event_slug = recent_event["slug"]

bouts_response = requests.get(
    f"https://api.citoapi.com/api/v1/ufc/events/{event_slug}/bouts",
    headers={"x-api-key": cito_api_key},
    timeout=30,
)

print("HTTP status:", bouts_response.status_code)

bouts_data = bouts_response.json()

HTTP status: 200


In [104]:
for bout in bouts_data["data"]:
    fighter_names = [
        fighter["fighterName"]
        for fighter in bout["fighters"]
    ]

    print(
        bout["id"],
        "-",
        " vs ".join(fighter_names),
        "-",
        bout["method"],
    )

13008 - Jose Miguel Delgado vs Jean Silva - Submission
12977 - Joseph Morales vs Brandon Moreno - Decision - Split
fight-night-september-12-2026-marwan-rahiki-tommy-mcmillen - Marwan Rahiki vs Tommy McMillen - Decision - Unanimous
12976 - Alexa Grasso vs Manon Fiorot - Decision - Unanimous
12975 - Curtis Blaydes vs Waldo Cortes Acosta - Decision - Unanimous
13000 - Dan Ige vs David Martinez - Decision - Unanimous
12978 - Edgar Chairez vs Tim Elliott - Decision - Unanimous
13001 - Muslim Salikhov vs Ignacio Bahamondes - Decision - Unanimous
12979 - Djorden Santos vs Yousri Belgaroui - KO/TKO
13118 - Tommy Gantt vs Drakkar Klose - Submission
fight-night-september-12-2026-rafa-garcia-zhu-rong - Rongzhu vs Rafa Garcia - Decision - Unanimous
13119 - Jessie Rosas vs Sean King III - KO/TKO
12980 - Regina Tarin vs JJ Aldrich - Decision - Unanimous


In [105]:
test_bout_id = "12975"

rounds_response = requests.get(
    f"https://api.citoapi.com/api/v1/ufc/bouts/{test_bout_id}/rounds",
    headers={"x-api-key": cito_api_key},
    timeout=30,
)

print("HTTP status:", rounds_response.status_code)

HTTP status: 200


In [106]:
rounds_data = rounds_response.json()

print(rounds_data.keys())
print(type(rounds_data["data"]))
print("Rows returned:", len(rounds_data["data"]))

dict_keys(['success', 'data', 'meta'])
<class 'list'>
Rows returned: 6


In [107]:
rounds_data["data"][0]

{'id': 'e5dfafc4-b92e-40fc-b955-ebeb0f310cff',
 'boutId': '12975',
 'fighterSlug': 'curtis-blaydes',
 'fighterName': 'Curtis Blaydes',
 'round': 1,
 'knockdowns': 0,
 'significantStrikes': '7 of 27',
 'totalStrikes': '7 of 27',
 'takedowns': '1 of 3',
 'submissionAttempts': 0,
 'reversals': 0,
 'controlTime': '1:13',
 'head': '1 of 20',
 'body': '2 of 2',
 'leg': '4 of 5',
 'distance': '6 of 26',
 'clinch': '0 of 0',
 'ground': '1 of 1',
 'lastSyncedAt': '2026-09-12T22:57:42.270Z'}

In [108]:
round_stats_df = pd.DataFrame(rounds_data["data"])

round_stats_df[
    [
        "fighterName",
        "round",
        "knockdowns",
        "significantStrikes",
        "totalStrikes",
        "takedowns",
        "submissionAttempts",
        "controlTime",
        "head",
        "body",
        "leg",
        "distance",
        "clinch",
        "ground",
    ]
]

,fighterName,round,knockdowns,significantStrikes,totalStrikes,takedowns,submissionAttempts,controlTime,head,body,leg,distance,clinch,ground
0,Curtis Blaydes,1,0,7 of 27,7 of 27,1 of 3,0,1:13,1 of 20,2 of 2,4 of 5,6 of 26,0 of 0,1 of 1
1,Waldo Cortes Acosta,1,0,7 of 23,8 of 24,0 of 0,0,0:00,5 of 20,0 of 0,2 of 3,7 of 23,0 of 0,0 of 0
2,Curtis Blaydes,2,0,9 of 24,16 of 31,2 of 3,0,1:27,3 of 18,1 of 1,5 of 5,9 of 24,0 of 0,0 of 0
3,Waldo Cortes Acosta,2,0,9 of 36,9 of 36,0 of 0,0,0:00,7 of 34,0 of 0,2 of 2,9 of 36,0 of 0,0 of 0
4,Curtis Blaydes,3,0,8 of 23,9 of 24,1 of 4,0,0:58,1 of 12,1 of 2,6 of 9,8 of 23,0 of 0,0 of 0
5,Waldo Cortes Acosta,3,0,12 of 36,13 of 37,0 of 0,0,0:00,9 of 33,0 of 0,3 of 3,12 of 36,0 of 0,0 of 0


In [109]:
def parse_landed_attempted(value):
    landed, attempted = value.split(" of ")
    return int(landed), int(attempted)


validation_rows = []

for _, row in round_stats_df.iterrows():
    sig_landed, sig_attempted = parse_landed_attempted(
        row["significantStrikes"]
    )

    head_landed, head_attempted = parse_landed_attempted(row["head"])
    body_landed, body_attempted = parse_landed_attempted(row["body"])
    leg_landed, leg_attempted = parse_landed_attempted(row["leg"])

    distance_landed, distance_attempted = parse_landed_attempted(
        row["distance"]
    )
    clinch_landed, clinch_attempted = parse_landed_attempted(row["clinch"])
    ground_landed, ground_attempted = parse_landed_attempted(row["ground"])

    validation_rows.append({
        "fighter": row["fighterName"],
        "round": row["round"],
        "target_landed_matches": (
            head_landed + body_landed + leg_landed == sig_landed
        ),
        "target_attempted_matches": (
            head_attempted + body_attempted + leg_attempted == sig_attempted
        ),
        "position_landed_matches": (
            distance_landed + clinch_landed + ground_landed == sig_landed
        ),
        "position_attempted_matches": (
            distance_attempted
            + clinch_attempted
            + ground_attempted
            == sig_attempted
        ),
    })

validation_df = pd.DataFrame(validation_rows)
validation_df

,fighter,round,target_landed_matches,target_attempted_matches,position_landed_matches,position_attempted_matches
0,Curtis Blaydes,1,True,True,True,True
1,Waldo Cortes Acosta,1,True,True,True,True
2,Curtis Blaydes,2,True,True,True,True
3,Waldo Cortes Acosta,2,True,True,True,True
4,Curtis Blaydes,3,True,True,True,True
5,Waldo Cortes Acosta,3,True,True,True,True


### Cito API Validation Conclusion

Cito was successfully queried from Python using an API key stored outside
the repository.

Initial validation found:

- fighter search returns structured fighter IDs and profile data
- Cito provides newer fighter information than the historical Kaggle snapshot
- overlapping fighter attributes generally agree across sources
- some career aggregates differ because the sources represent different points
  in time or use slightly different source values
- Cito exposes event, bout, and true round-level statistics
- historical round-stat access is restricted by API plan
- recent round-stat data was successfully retrieved
- a three-round fight returned six records: one fighter-round record per fighter
  per round
- head/body/leg and distance/clinch/ground breakdowns reconciled exactly with
  significant-strike totals for every tested round

Cito appears useful as a current-data and round-stat provider, but provider data
should still be validated and normalized before being used by UPSET.

In [110]:
import json

# Find the specific fight in the response we previously fetched.
raw_bout = next(
    bout
    for bout in bouts_data["data"]
    if str(bout["id"]) == "12975"
)

# Display every field, including the nested fighter information.
print(json.dumps(raw_bout, indent=2))

{
  "id": "12975",
  "eventSlug": "ufc-fight-night-september-12-2026",
  "cardSection": "Main Card",
  "cardSectionOrder": 1,
  "cardPosition": "Main Card 5",
  "boutOrder": 1005,
  "weightClass": "Heavyweight",
  "titleBout": false,
  "status": "completed",
  "isCancelled": false,
  "cancellationReason": null,
  "winnerFighterSlug": "curtis-blaydes",
  "resultRound": 3,
  "resultTime": "5:00",
  "method": "Decision - Unanimous",
  "methodDetails": null,
  "odds": {
    "red": null,
    "blue": null
  },
  "awards": null,
  "dataAvailability": {
    "result": "available",
    "warning": "Using cached data",
    "warnings": [],
    "fetchedAt": "2026-09-16T07:10:13.833Z",
    "roundStats": "when_available",
    "dataAgeHours": 14.15,
    "dataFreshness": "live",
    "fighterCorners": "available"
  },
  "createdAt": "2026-08-05T20:15:47.125Z",
  "lastSyncedAt": "2026-09-16T21:16:20.122Z",
  "fighters": [
    {
      "id": "d29440ca-edfb-4877-9acc-611b85ebc874",
      "boutId": "12975",
 

In [111]:
for index, fighter in enumerate(raw_bout["fighters"], start=1):
    print(f"\nFighter {index}")

    for key, value in fighter.items():
        if isinstance(value, (dict, list)):
            print(f"  {key}: [nested information]")
        else:
            print(f"  {key}: {value}")


Fighter 1
  id: d29440ca-edfb-4877-9acc-611b85ebc874
  boutId: 12975
  fighterSlug: curtis-blaydes
  fighterName: Curtis Blaydes
  corner: blue
  outcome: win
  rankText: None
  country: United States
  odds: None
  imageUrl: https://ufc.com/images/styles/event_fight_card_upper_body_of_standing_athlete/s3/2026-09/BLAYDES_CURTIS_R_09-12.png?itok=hrU5rCtK
  fighterId: ffbb498d-8144-48d7-bd5b-dc02b94c9e6e
  flag: 🇺🇸
  championStatus: none
  profile: [nested information]

Fighter 2
  id: b43735b4-1329-4b5a-817a-2bf09cc58e31
  boutId: 12975
  fighterSlug: waldo-cortes-acosta
  fighterName: Waldo Cortes Acosta
  corner: red
  outcome: loss
  rankText: None
  country: Dominican Republic
  odds: None
  imageUrl: https://ufc.com/images/styles/event_fight_card_upper_body_of_standing_athlete/s3/2026-09/CORTES_ACOSTA_WALDO_L_09-12.png?itok=xe7OyefZ
  fighterId: ae6aa622-1b62-4cac-ab64-9179aa9f07af
  flag: 🇩🇴
  championStatus: none
  profile: [nested information]


In [112]:
import importlib

from upset.data import normalization

# Reload so this notebook sees the function you just saved.
importlib.reload(normalization)

fight = normalization.normalize_cito_fight(raw_bout)

print("Bout ID:", fight.source_bout_id)
print("Fighter 1:", fight.fighter_1_name)
print("Fighter 1 ID:", fight.source_fighter_1_id)
print("Fighter 2:", fight.fighter_2_name)
print("Fighter 2 ID:", fight.source_fighter_2_id)
print("Winner:", fight.winner_name)
print("Method:", fight.result_method)
print("Round:", fight.result_round)
print("Time:", fight.result_time)

Bout ID: 12975
Fighter 1: Curtis Blaydes
Fighter 1 ID: ffbb498d-8144-48d7-bd5b-dc02b94c9e6e
Fighter 2: Waldo Cortes Acosta
Fighter 2 ID: ae6aa622-1b62-4cac-ab64-9179aa9f07af
Winner: Curtis Blaydes
Method: Decision - Unanimous
Round: 3
Time: 5:00


In [115]:
# Find the event whose slug matches the fight's eventSlug.
raw_event = next(
    event_record
    for event_record in recent_data["data"]
    if event_record["slug"] == raw_bout["eventSlug"]
)

# Convert the event into UPSET's format.
event = normalization.normalize_cito_event(raw_event)

# Convert the fight and attach the verified event ID.
fight = normalization.normalize_cito_fight(raw_bout, event=event)

print("Event:", event.title)
print("Date:", event.event_date)
print("Event ID:", event.source_event_id)
print("Fight's event ID:", fight.source_event_id)

assert fight.source_event_id == event.source_event_id
print("Event connection verified.")

Event: Noche UFC
Date: 2026-09-12
Event ID: af7fb97f-bc2a-440b-abdc-b03d4228bf44
Fight's event ID: af7fb97f-bc2a-440b-abdc-b03d4228bf44
Event connection verified.
